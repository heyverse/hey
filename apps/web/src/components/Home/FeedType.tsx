import New from "@components/Shared/Badges/New";
import { HomeFeedType } from "@hey/data/enums";
import { TabButton } from "@hey/ui";
import type { FC, JSX } from "react";
import { useHomeTabStore } from "src/store/persisted/useHomeTabStore";

const FeedType: FC = () => {
  const { feedType, setFeedType } = useHomeTabStore();

  const tabs = [
    { name: "Following", type: HomeFeedType.FOLLOWING },
    {
      badge: <New />,
      name: "For You",
      type: HomeFeedType.FORYOU
    }
  ].filter(
    (
      tab
    ): tab is {
      badge?: JSX.Element;
      name: string;
      type: HomeFeedType;
    } => Boolean(tab)
  );

  return (
    <div className="flex flex-wrap gap-3 px-5 sm:px-0">
      {tabs.map((tab) => (
        <TabButton
          active={feedType === tab.type}
          badge={tab.badge}
          key={tab.type}
          name={tab.name}
          onClick={() => setFeedType(tab.type)}
          showOnSm
        />
      ))}
    </div>
  );
};

export default FeedType;
