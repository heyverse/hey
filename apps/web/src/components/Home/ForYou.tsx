import QueuedPost from "@components/Post/QueuedPost";
import SinglePost from "@components/Post/SinglePost";
import PostsShimmer from "@components/Shared/Shimmer/PostsShimmer";
import { LightBulbIcon } from "@heroicons/react/24/outline";
import {
  type MlpostsForYouRequest,
  PageSize,
  type Post,
  useMlPostsForYouQuery
} from "@hey/indexer";
import { OptimisticTxType } from "@hey/types/enums";
import { Card, EmptyState, ErrorMessage } from "@hey/ui";
import type { FC } from "react";
import { Virtuoso } from "react-virtuoso";
import { useAccountStore } from "src/store/persisted/useAccountStore";
import { useTransactionStore } from "src/store/persisted/useTransactionStore";

const ForYou: FC = () => {
  const { currentAccount } = useAccountStore();
  const { txnQueue } = useTransactionStore();

  const request: MlpostsForYouRequest = {
    pageSize: PageSize.Fifty,
    account: currentAccount?.address,
    shuffle: true
  };

  const { data, error, fetchMore, loading } = useMlPostsForYouQuery({
    variables: { request }
  });

  const posts = data?.mlPostsForYou.items;
  const pageInfo = data?.mlPostsForYou.pageInfo;
  const hasMore = pageInfo?.next;

  const onEndReached = async () => {
    if (hasMore) {
      await fetchMore({
        variables: { request: { ...request, cursor: pageInfo?.next } }
      });
    }
  };

  if (loading) {
    return <PostsShimmer />;
  }

  if (posts?.length === 0) {
    return (
      <EmptyState
        icon={<LightBulbIcon className="size-8" />}
        message="No posts yet!"
      />
    );
  }

  if (error) {
    return <ErrorMessage error={error} title="Failed to load highlights" />;
  }

  return (
    <>
      {txnQueue.map((txn) =>
        txn?.type === OptimisticTxType.CREATE_POST ? (
          <QueuedPost key={txn.txHash} txn={txn} />
        ) : null
      )}
      <Card>
        <Virtuoso
          className="virtual-divider-list-window"
          computeItemKey={(index, item) => `${item.post.id}-${index}`}
          data={posts}
          endReached={onEndReached}
          itemContent={(index, item) => (
            <SinglePost
              isFirst={index === 0}
              isLast={index === (posts?.length || 0) - 1}
              post={item.post as Post}
            />
          )}
          useWindowScroll
        />
      </Card>
    </>
  );
};

export default ForYou;
