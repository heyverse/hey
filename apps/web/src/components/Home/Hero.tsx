import { STATIC_IMAGES_URL } from "@hey/data/constants";

const Hero = () => {
  return (
    <div
      className="relative h-64 w-full gap-y-5 rounded-none bg-black md:rounded-xl"
      style={{
        backgroundImage: `url(${STATIC_IMAGES_URL}/hero.webp)`,
        backgroundPosition: "center",
        backgroundSize: "cover"
      }}
    >
      <div className="absolute bottom-5 left-5">
        <div className="font-extrabold text-3xl text-white">Welcome to Hey</div>
        <div className="font-extrabold text-gray-200">
          a social network built on Lens
        </div>
      </div>
    </div>
  );
};

export default Hero;
