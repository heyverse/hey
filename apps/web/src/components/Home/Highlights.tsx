import SinglePost from "@/components/Post/SinglePost";
import PostsShimmer from "@/components/Shared/Shimmer/PostsShimmer";
import { Card, EmptyState, ErrorMessage } from "@/components/Shared/UI";
import { useAccountStore } from "@/store/persisted/useAccountStore";
import { LightBulbIcon } from "@heroicons/react/24/outline";
import {
  PageSize,
  type TimelineHighlightsRequest,
  useTimelineHighlightsQuery
} from "@hey/indexer";
import { useIntersectionObserver } from "@uidotdev/usehooks";
import { useEffect } from "react";
import { WindowVirtualizer } from "virtua";

const Highlights = () => {
  const { currentAccount } = useAccountStore();
  const [ref, entry] = useIntersectionObserver({
    threshold: 0,
    root: null,
    rootMargin: "0px"
  });

  const request: TimelineHighlightsRequest = {
    pageSize: PageSize.Fifty,
    account: currentAccount?.address
  };

  const { data, error, fetchMore, loading } = useTimelineHighlightsQuery({
    variables: { request }
  });

  const posts = data?.timelineHighlights.items;
  const pageInfo = data?.timelineHighlights.pageInfo;
  const hasMore = pageInfo?.next;

  const onEndReached = async () => {
    if (hasMore) {
      await fetchMore({
        variables: { request: { ...request, cursor: pageInfo?.next } }
      });
    }
  };

  useEffect(() => {
    if (entry?.isIntersecting) {
      onEndReached();
    }
  }, [entry?.isIntersecting]);

  if (loading) {
    return <PostsShimmer />;
  }

  if (!posts?.length) {
    return (
      <EmptyState
        icon={<LightBulbIcon className="size-8" />}
        message="No posts yet!"
      />
    );
  }

  if (error) {
    return <ErrorMessage error={error} title="Failed to load highlights" />;
  }

  const filteredPosts = posts.filter(
    (post) =>
      !post.author.operations?.hasBlockedMe &&
      !post.author.operations?.isBlockedByMe &&
      !post.operations?.hasReported
  );

  return (
    <Card className="virtual-divider-list-window">
      <WindowVirtualizer>
        {filteredPosts.map((post) => (
          <SinglePost key={post.id} post={post} />
        ))}
        {hasMore && <span ref={ref} />}
      </WindowVirtualizer>
    </Card>
  );
};

export default Highlights;
