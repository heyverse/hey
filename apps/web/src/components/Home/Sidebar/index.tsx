import SignupCard from "@components/Shared/Auth/SignupCard";
import Footer from "@components/Shared/Footer";
import type { FC } from "react";
import { memo } from "react";
import { useAccountStore } from "src/store/persisted/useAccountStore";
import SetAccount from "./SetAccount";
import StaffPicks from "./StaffPicks";
import WhoToFollow from "./WhoToFollow";

const Sidebar: FC = () => {
  const { currentAccount } = useAccountStore();
  const loggedInWithProfile = Boolean(currentAccount);
  const loggedOut = !loggedInWithProfile;

  return (
    <>
      {/* <Gitcoin /> */}
      {loggedOut && <SignupCard />} {/* Onboarding steps */}
      {loggedInWithProfile && <SetAccount />}
      {/* Recommendations */}
      <StaffPicks />
      {loggedInWithProfile && <WhoToFollow />}
      <Footer />
    </>
  );
};

export default memo(Sidebar);
