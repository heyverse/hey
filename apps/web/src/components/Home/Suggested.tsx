import DismissRecommendedAccount from "@components/Shared/DismissRecommendedAccount";
import SingleAccount from "@components/Shared/SingleAccount";
import { UsersIcon } from "@heroicons/react/24/outline";
import type { AccountFragment } from "@hey/indexer";
import { EmptyState } from "@hey/ui";
import type { FC } from "react";
import { Virtuoso } from "react-virtuoso";
import { useAccountStore } from "src/store/persisted/useAccountStore";

interface SuggestedProps {
  accounts: AccountFragment[];
}

const Suggested: FC<SuggestedProps> = ({ accounts }) => {
  const { currentAccount } = useAccountStore();

  if (accounts.length === 0) {
    return (
      <EmptyState
        hideCard
        icon={<UsersIcon className="size-8" />}
        message="Nothing to suggest"
      />
    );
  }

  return (
    <div className="max-h-[80vh] overflow-y-auto">
      <Virtuoso
        className="virtual-account-list"
        // remove the first 5 profiles from the list because they are already shown in the sidebar
        data={accounts.slice(5)}
        itemContent={(_, account) => (
          <div className="flex items-center space-x-3 p-5">
            <div className="w-full">
              <SingleAccount
                hideFollowButton={currentAccount?.address === account.address}
                hideUnfollowButton={currentAccount?.address === account.address}
                account={account}
                showBio
                showUserPreview={false}
              />
            </div>
            <DismissRecommendedAccount account={account} />
          </div>
        )}
      />
    </div>
  );
};

export default Suggested;
