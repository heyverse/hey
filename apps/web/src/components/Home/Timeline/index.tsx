import QueuedPost from "@components/Post/QueuedPost";
import SinglePost from "@components/Post/SinglePost";
import PostsShimmer from "@components/Shared/Shimmer/PostsShimmer";
import { UserGroupIcon } from "@heroicons/react/24/outline";
import {
  type Post,
  type TimelineItem,
  type TimelineRequest,
  useTimelineQuery
} from "@hey/indexer";
import { OptimisticTxType } from "@hey/types/enums";
import { Card, EmptyState, ErrorMessage } from "@hey/ui";
import type { FC } from "react";
import { memo, useRef } from "react";
import type { StateSnapshot, VirtuosoHandle } from "react-virtuoso";
import { Virtuoso } from "react-virtuoso";
import { useAccountStore } from "src/store/persisted/useAccountStore";
import { useTransactionStore } from "src/store/persisted/useTransactionStore";

let virtuosoState: any = { ranges: [], screenTop: 0 };

const Timeline: FC = () => {
  const { currentAccount } = useAccountStore();
  const { txnQueue } = useTransactionStore();
  const virtuoso = useRef<VirtuosoHandle>(null);

  const request: TimelineRequest = {
    account: currentAccount?.address
  };

  const { data, error, fetchMore, loading } = useTimelineQuery({
    fetchPolicy: "cache-and-network",
    variables: { request }
  });

  const feed = data?.timeline?.items;
  const pageInfo = data?.timeline?.pageInfo;
  const hasMore = pageInfo?.next;

  const onScrolling = (scrolling: boolean) => {
    if (!scrolling) {
      virtuoso?.current?.getState((state: StateSnapshot) => {
        virtuosoState = { ...state };
      });
    }
  };

  const onEndReached = async () => {
    if (hasMore) {
      await fetchMore({
        variables: { request: { ...request, cursor: pageInfo?.next } }
      });
    }
  };

  if (loading) {
    return <PostsShimmer />;
  }

  if (feed?.length === 0) {
    return (
      <EmptyState
        icon={<UserGroupIcon className="size-8" />}
        message="No posts yet!"
      />
    );
  }

  if (error) {
    return <ErrorMessage error={error} title="Failed to load timeline" />;
  }

  return (
    <>
      {txnQueue.map((txn) =>
        txn?.type !== OptimisticTxType.CREATE_COMMENT ? (
          <QueuedPost key={txn.txHash} txn={txn} />
        ) : null
      )}
      <Card>
        <Virtuoso
          className="virtual-divider-list-window"
          computeItemKey={(index, timelineItem) =>
            `${timelineItem.id}-${index}`
          }
          data={feed}
          endReached={onEndReached}
          isScrolling={onScrolling}
          itemContent={(index, timelineItem) => (
            <SinglePost
              timelineItem={timelineItem as TimelineItem}
              isFirst={index === 0}
              isLast={index === (feed?.length || 0) - 1}
              post={timelineItem.primary as Post}
            />
          )}
          ref={virtuoso}
          restoreStateFrom={
            virtuosoState.ranges.length === 0
              ? virtuosoState?.current?.getState(
                  (state: StateSnapshot) => state
                )
              : virtuosoState
          }
          useWindowScroll
        />
      </Card>
    </>
  );
};

export default memo(Timeline);
