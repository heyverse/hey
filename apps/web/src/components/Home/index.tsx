import NewPost from "@components/Composer/NewPost";
import ExploreFeed from "@components/Explore/ExploreFeed";
import { HomeFeedType } from "@hey/data/enums";
import { GridItemEight, GridItemFour, GridLayout } from "@hey/ui";
import type { NextPage } from "next";
import { useAccountStore } from "src/store/persisted/useAccountStore";
import { useHomeTabStore } from "src/store/persisted/useHomeTabStore";
import FeedType from "./FeedType";
import ForYou from "./ForYou";
import Hero from "./Hero";
import Sidebar from "./Sidebar";
import Timeline from "./Timeline";

const Home: NextPage = () => {
  const { currentAccount } = useAccountStore();
  const { feedType } = useHomeTabStore();

  const loggedInWithProfile = Boolean(currentAccount);

  return (
    <>
      {!loggedInWithProfile && <Hero />}
      <GridLayout>
        <GridItemEight className="space-y-5">
          {loggedInWithProfile ? (
            <>
              <NewPost />
              <FeedType />
              {feedType === HomeFeedType.FOLLOWING ? (
                <Timeline />
              ) : feedType === HomeFeedType.FORYOU ? (
                <ForYou />
              ) : null}
            </>
          ) : (
            <ExploreFeed />
          )}
        </GridItemEight>
        <GridItemFour>
          <Sidebar />
        </GridItemFour>
      </GridLayout>
    </>
  );
};

export default Home;
