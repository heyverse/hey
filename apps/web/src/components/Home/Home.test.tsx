import { render } from "@testing-library/react";
import type React from "react";
import { describe, expect, it, vi } from "vitest";
import Home from "./index";

vi.mock("../Composer/NewPost", () => ({ default: () => <div>NewPost</div> }));
vi.mock("../Explore/ExploreFeed", () => ({
  default: () => <div>ExploreFeed</div>
}));
vi.mock("../Shared/PageLayout", () => ({
  default: ({ children }: { children: React.ReactNode }) => (
    <div>{children}</div>
  )
}));
vi.mock("../../store/persisted/useAccountStore", () => ({
  useAccountStore: () => ({ currentAccount: undefined })
}));
vi.mock("../../store/persisted/useHomeTabStore", () => ({
  useHomeTabStore: () => ({ feedType: "FORYOU" })
}));
vi.mock("./FeedType", () => ({ default: () => <div>FeedType</div> }));
vi.mock("./ForYou", () => ({ default: () => <div>ForYou</div> }));
vi.mock("./Hero", () => ({ default: () => <div>Hero</div> }));
vi.mock("./Highlights", () => ({ default: () => <div>Highlights</div> }));
vi.mock("./Timeline", () => ({ default: () => <div>Timeline</div> }));

describe("Home", () => {
  it("renders correctly", () => {
    const { container } = render(<Home />);
    expect(container).toMatchSnapshot();
  });
});
