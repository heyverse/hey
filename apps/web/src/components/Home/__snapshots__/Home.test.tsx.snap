// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home > renders correctly 1`] = `
<div>
  <div>
    <div>
      Hero
    </div>
    <div>
      ExploreFeed
    </div>
  </div>
</div>
`;
